lib/switchtec.c

```diff
@@ -592,8 +618,8 @@ int switchtec_log_to_file(struct switchtec_dev *dev,
 		return log_b_to_file(dev, MRPC_FWLOGRD_THRD, fd);
 	};
 
-	errno = EINVAL;
-	return -errno;
+	set_errno(EINVAL);
```

Comment:
This seems like it should be in separate patches. If it's even necessary which I'm not convinced of. There's a large number of places that set errno and we have to be careful with this.

My expectation is seeing 'mrpc' is only set when an mrpc command fails we should be mostly fine seeing all the existing generally stops when there is an error.  However, we may need a clear error function to reset mrpc to -1 when someone sets errno to 0.

Reply:
I considered these two solutions. The PROs of current one is that for upcoming APIs we won't need to worry about to reset the mrpc to -1, and we also do not have to stop after an error. The Cons is we need to go through all the places that set errno.

If we introduce another error function to clear error, we will need to be very careful to figure out if the path prior to the error has updated 'mrpc' or not. I think although we have two variables 'mrpc' and 'errno', we should treat them as one, and the operation should be atomic. It would make the concept more clear and less restriciton in the coding.

I'd think that going through all the places that set errno is the cost we need to pay.

Reply:
Well all our APIs support not stopping, it's just that we don't have much code that does this.

It's not that hard to go through all the cases where we set the errno. Except the problem is you can't change all the other system calls that do this. So if the user gets a failed MRPC command followed by a failed malloc (for example) then it will look like an MRPC error when they call switchtec_perror().

The downside of the clear errno function is that users of our libraries probably won't realise they need to do this. But I think it's the only correct way to do it even if it's hard on the users.

Another thought I had was to use a couple high bits in errno to indicate that the mrpc_error is valid. That's probably the best choice unless I'm missing something. That way when errno is cleared we know to ignore mrpc_error.

Reply:
> So if the user gets a failed MRPC command followed by a failed malloc (for example) then it will look like an MRPC error when they call switchtec_perror().

I'm thinking it's a inproper use of our APIs. User should call switchtec_perror() immediately after a switchtec API. And for a system call error output, user should use the perror() instead of swtichtec_perror(). We only need to ensure that the error of each switchtec API function be properly returned.
```
ret = switchtec_XXX();
if (ret)
        switchtec_perror();
p = malloc();
if (p == NULL)
        switchtec_perror(); --> perror();
```
> Another thought I had was to use a couple high bits in errno to indicate that the mrpc_error is valid. 

Because the 'MRPC Command Return Value' takes 4 bytes, and the errno is of type int, I don't think it's a good idea to encode anything into errno, which may break the MRPC interface.

Reply:
No, that doesn't change anything. What if instead of malloc() they call one of our API's that fails because of an internal malloc() call?  We simply can not go through all our code and clear mrpc_error for every system call that might set errno.

So I still think that using a single high bit to say that the mrpc_errno is the correct approach. That way anything else that overwrites errno automatically clears the bit and then we don't check mrpc_errno. I don't think there's any error code that uses more than 30 bits so using the 31st bit (with the 32nd bit as the sign) should be acceptable.

Reply:
Unluckily, below error code is using all bits. Even though we can mask off the high bit in a MRPC call, as it's the only one, but there's still chance firmware add other error code like this to make this workaround invalid. 

`ERR_ACCESS_REFUSED = 0xFFFF0001`

If we want to ensure each switchtec API funciton returns (mrpc_error_cmd, errno) pair correctly, we don't actually need to go through every system call, but just clear the errno after a non stopped switchtec function call in the switchtec function implementation (for a void returning switchtec function, we may need to do the same thing).  So the convention turns out to be that if a switchtec function calls another switchtec function, either stop on error or set_errno(0) (mrpc_error_cmd = -1, errno = 0), if If all switchtec functions return (mrpc_error_cmd, errno) correctly, users do not need to clear errno themselves. What do you think? 
```
int switchtec_functionA()
{
        switchtec_funtionB();
        set_errno(0);        // mrpc_error_cmd = -1, errno = 0
        p = malloc();
        if (p == NULL)
                return -1;
}
```

Reply:
No, user code need to follow the same convention.

```
switchtec_functionA();
set_errno(0);
switchtec_functionB();
```

Reply:
The point is that adding this set_error() function to the api requirements is complicated and you can't expect your callers to get it correct (or even us to get it correct inside the library). It'll take a bunch of nasty code audits to ensure everything is right.

Allowing for the last setter of errno to override the mrpc_errno is a huge win in terms of the complexity we have to consider when using errno.

I don't buy that we have to plan for the firmware to set the high bit for three reasons:

1) If the firmware happens to set the high bit it's fine because we already need it set.

2) We have control over the firmware. Go tell them to limit the error number to 30 bits or push back if they add an error code that high. 30 bits is more than enough for any of their future purposes.

3) If we absolutely had to, we could map the crazy error with the high bit set into a sane lower numbered error.

The only concern is if some upper layer software wants to use that bit for something. This is pretty rare though and in this case using switchtec_perror() would be invalid.

Reply:
I've aligned with firmware, they agreed to remove the 0xffff0001 return code from Gen4, and make sure the highest two bits are not used (cleared). I'll update with the bit 31st as the flag and keep 32nd bit zero as the sign.